File: pacman.go
// pacman.go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"time"
	"github.com/nsf/termbox-go"
)

const recordFile = ".pacman_record.json"
type Record struct{ Best int }

func loadRecord() int {
	f, err := os.Open(recordFile)
	if err != nil { return 0 }
	defer f.Close()
	var r Record
	json.NewDecoder(f).Decode(&r)
	return r.Best
}
func saveRecord(best int) {
	f, _ := os.Create(recordFile)
	defer f.Close()
	json.NewEncoder(f).Encode(Record{best})
}

func main() {
	speed := 150
	if len(os.Args) > 2 && os.Args[1] == "-s" {
		if s, err := strconv.Atoi(os.Args[2]); err == nil && s > 0 { speed = s }
	}
	err := termbox.Init()
	if err != nil { fmt.Println(err); return }
	defer termbox.Close()
	termbox.SetInputMode(termbox.InputEsc)
	w, h := termbox.Size()
	if h < 22 || w < 42 { fmt.Println("Terminal too small"); return }
	rand.Seed(time.Now().UnixNano())

	const H, W = 20, 20
	mapData := [][]int{
		{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
		{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
		{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
		{0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0},
		{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
		{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
		{0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0},
		{0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0},
		{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
		{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
		{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
		{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
		{0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0},
		{0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0},
		{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
		{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
		{0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0},
		{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
		{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
		{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
	}
	pacX, pacY := 9, 9
	dirX, dirY := 0, 0
	score, lives, level := 0, 3, 1
	best := loadRecord()
	gameOver := false
	frame := time.Duration(speed) * time.Millisecond

	type Ghost struct{ x, y, color, hx, hy int; scared bool }
	ghosts := []Ghost{
		{9,8,2,9,8,false},
		{8,9,2,8,9,false},
		{10,9,2,10,9,false},
		{9,10,2,9,10,false},
	}
	scaredTimer := 0
	totalDots := func() int {
		c := 0
		for _, row := range mapData {
			for _, v := range row {
				if v == 1 || v == 2 { c++ }
			}
		}
		return c
	}
	total := totalDots()

	tbprint := func(x, y int, fg, bg termbox.Attribute, msg string) {
		for _, ch := range msg {
			termbox.SetCell(x, y, ch, fg, bg)
			x++
		}
	}

	for {
		ev := termbox.PollEvent()
		if ev.Type == termbox.EventKey {
			if ev.Key == termbox.KeyEsc || ev.Ch == 'q' { return }
			if ev.Ch == 'r' && gameOver {
				// reset
				mapData = [][]int{
					{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
					{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
					{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
					{0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0},
					{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
					{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
					{0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0},
					{0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0},
					{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
					{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
					{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
					{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
					{0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0},
					{0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0},
					{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
					{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
					{0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0},
					{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
					{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
					{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
				}
				pacX, pacY = 9, 9
				dirX, dirY = 0, 0
				score, lives, level = 0, 3, 1
				gameOver = false
				scaredTimer = 0
				for i := range ghosts {
					ghosts[i].x = ghosts[i].hx
					ghosts[i].y = ghosts[i].hy
					ghosts[i].scared = false
				}
				total = totalDots()
				continue
			}
			if ev.Key == termbox.KeyArrowLeft || ev.Ch == 'a' { dirX, dirY = 0, -1 }
			if ev.Key == termbox.KeyArrowRight || ev.Ch == 'd' { dirX, dirY = 0, 1 }
			if ev.Key == termbox.KeyArrowUp || ev.Ch == 'w' { dirX, dirY = -1, 0 }
			if ev.Key == termbox.KeyArrowDown || ev.Ch == 's' { dirX, dirY = 1, 0 }
		}

		if gameOver {
			termbox.Clear(termbox.ColorDefault, termbox.ColorDefault)
			msg := fmt.Sprintf("GAME OVER! Score: %d  Best: %d", score, best)
			tbprint(w/2-len(msg)/2, h/2-2, termbox.ColorWhite, termbox.ColorDefault, msg)
			tbprint(w/2-15, h/2, termbox.ColorCyan, termbox.ColorDefault, "R - restart | Q - quit")
			termbox.Flush()
			continue
		}

		// Move Pac-Man
		ny, nx := pacY+dirX, pacX+dirY
		if ny >= 0 && ny < H && nx >= 0 && nx < W && mapData[ny][nx] != 0 {
			pacY, pacX = ny, nx
			if mapData[ny][nx] == 1 { score += 10; mapData[ny][nx] = 3; total-- }
			if mapData[ny][nx] == 2 { score += 50; mapData[ny][nx] = 3; total--; scaredTimer = 30; for i := range ghosts { ghosts[i].scared = true } }
		}

		// Move ghosts
		for i := range ghosts {
			g := &ghosts[i]
			dirs := [][2]int{{1,0},{-1,0},{0,1},{0,-1}}
			if g.scared {
				// убегаем – случайно
				idx := rand.Intn(len(dirs))
				dx, dy := dirs[idx][0], dirs[idx][1]
				nx2, ny2 := g.x+dx, g.y+dy
				if nx2 >= 0 && nx2 < W && ny2 >= 0 && ny2 < H && mapData[ny2][nx2] != 0 {
					g.x, g.y = nx2, ny2
				}
			} else {
				idx := rand.Intn(len(dirs))
				dx, dy := dirs[idx][0], dirs[idx][1]
				nx2, ny2 := g.x+dx, g.y+dy
				if nx2 >= 0 && nx2 < W && ny2 >= 0 && ny2 < H && mapData[ny2][nx2] != 0 {
					g.x, g.y = nx2, ny2
				}
			}
		}

		// Collisions
		for i := range ghosts {
			g := &ghosts[i]
			if g.x == pacX && g.y == pacY {
				if g.scared {
					score += 100
					g.x, g.y = g.hx, g.hy
					g.scared = false
				} else {
					lives--
					if lives <= 0 {
						gameOver = true
						if score > best { best = score; saveRecord(best) }
					} else {
						pacX, pacY = 9, 9
						for j := range ghosts {
							ghosts[j].x = ghosts[j].hx
							ghosts[j].y = ghosts[j].hy
							ghosts[j].scared = false
						}
					}
				}
				break
			}
		}

		if scaredTimer > 0 {
			scaredTimer--
			if scaredTimer == 0 {
				for i := range ghosts {
					ghosts[i].scared = false
				}
			}
		}

		if total == 0 {
			level++
			// reset map
			mapData = [][]int{
				{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
				{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
				{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
				{0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0},
				{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
				{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
				{0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0},
				{0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0},
				{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
				{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
				{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
				{0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0},
				{0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0},
				{0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0},
				{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
				{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
				{0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0},
				{0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0},
				{0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0},
				{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
			}
			total = totalDots()
			pacX, pacY = 9, 9
			for i := range ghosts {
				ghosts[i].x = ghosts[i].hx
				ghosts[i].y = ghosts[i].hy
				ghosts[i].scared = false
			}
		}

		// Draw
		termbox.Clear(termbox.ColorDefault, termbox.ColorDefault)
		for y := 0; y < H; y++ {
			for x := 0; x < W; x++ {
				v := mapData[y][x]
				if v == 0 {
					termbox.SetCell(x*2+1, y+1, '#', termbox.ColorWhite, termbox.ColorDefault)
				} else if v == 1 {
					termbox.SetCell(x*2+1, y+1, '.', termbox.ColorGreen, termbox.ColorDefault)
				} else if v == 2 {
					termbox.SetCell(x*2+1, y+1, 'O', termbox.ColorCyan, termbox.ColorDefault)
				}
			}
		}
		termbox.SetCell(pacX*2+1, pacY+1, 'C', termbox.ColorYellow|termbox.AttrBold, termbox.ColorDefault)
		for _, g := range ghosts {
			color := termbox.ColorBlue
			if !g.scared {
				color = termbox.ColorRed
			}
			termbox.SetCell(g.x*2+1, g.y+1, 'G', color|termbox.AttrBold, termbox.ColorDefault)
		}
		tbprint(2, 0, termbox.ColorWhite, termbox.ColorDefault, fmt.Sprintf("Score: %d", score))
		tbprint(W*2-12, 0, termbox.ColorWhite, termbox.ColorDefault, fmt.Sprintf("Best: %d", best))
		tbprint(W*2-25, 0, termbox.ColorWhite, termbox.ColorDefault, fmt.Sprintf("Lives: %d  Level: %d", lives, level))
		if gameOver {
			msg := "GAME OVER! Press R to restart, Q to quit"
			tbprint((W*2-len(msg))/2, H/2+1, termbox.ColorWhite, termbox.ColorDefault, msg)
		}
		termbox.Flush()
		time.Sleep(frame)
	}
}

File: pacman.py
# pacman.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys, os, random, json, time, argparse, curses
from pathlib import Path

RECORD_FILE = Path.home() / '.pacman_record.json'

def load_record():
    try:
        with open(RECORD_FILE) as f:
            return json.load(f).get('record', 0)
    except:
        return 0

def save_record(record):
    with open(RECORD_FILE, 'w') as f:
        json.dump({'record': record}, f)

# Карта: 0-стена, 1-точка, 2-суперточка, 3-пусто
MAP = [
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0],
    [0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0],
    [0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0],
    [0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0],
    [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0],
    [0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0],
    [0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0],
    [0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0],
    [0,1,1,1,1,0,0,1,0,0,0,0,1,0,0,1,1,1,1,0],
    [0,1,0,0,1,0,0,1,0,0,0,0,1,0,0,1,0,0,1,0],
    [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0],
    [0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0],
    [0,2,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,2,0],
    [0,1,0,1,0,0,1,0,0,1,1,0,0,1,0,0,1,0,1,0],
    [0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
]

def main(stdscr, speed):
    curses.curs_set(0)
    stdscr.nodelay(1)
    stdscr.timeout(0)
    curses.start_color()
    curses.use_default_colors()
    curses.init_pair(1, curses.COLOR_YELLOW, -1)   # pacman
    curses.init_pair(2, curses.COLOR_RED, -1)     # ghost
    curses.init_pair(3, curses.COLOR_GREEN, -1)   # points
    curses.init_pair(4, curses.COLOR_CYAN, -1)    # super points
    curses.init_pair(5, curses.COLOR_WHITE, -1)   # walls
    curses.init_pair(6, curses.COLOR_BLUE, -1)    # scared ghost

    height, width = stdscr.getmaxyx()
    if height < 22 or width < 42:
        print("Терминал слишком мал (нужно 22x42)")
        return

    H, W = 20, 20
    map_data = [row[:] for row in MAP]
    pac_x, pac_y = 9, 9
    pac_dir = (0, 0)
    score = 0
    lives = 3
    level = 1
    best = load_record()
    game_over = False
    frame = speed / 1000.0

    ghosts = [
        {'x': 9, 'y': 8, 'color': 2, 'dir': (1,0), 'scared': False, 'home': (9,8)},
        {'x': 8, 'y': 9, 'color': 2, 'dir': (0,1), 'scared': False, 'home': (8,9)},
        {'x': 10, 'y': 9, 'color': 2, 'dir': (0,-1), 'scared': False, 'home': (10,9)},
        {'x': 9, 'y': 10, 'color': 2, 'dir': (-1,0), 'scared': False, 'home': (9,10)},
    ]
    scared_timer = 0
    total_dots = sum(row.count(1) + row.count(2) for row in map_data)

    def draw():
        stdscr.clear()
        for y in range(H):
            for x in range(W):
                ch = map_data[y][x]
                if ch == 0:
                    stdscr.addch(y+1, x*2+1, '#', curses.color_pair(5))
                elif ch == 1:
                    stdscr.addch(y+1, x*2+1, '.', curses.color_pair(3))
                elif ch == 2:
                    stdscr.addch(y+1, x*2+1, 'O', curses.color_pair(4))
        # Pac-Man
        stdscr.addch(pac_y+1, pac_x*2+1, 'C', curses.color_pair(1) | curses.A_BOLD)
        # Ghosts
        for g in ghosts:
            color = curses.color_pair(6) if g['scared'] else curses.color_pair(g['color'])
            stdscr.addch(g['y']+1, g['x']*2+1, 'G', color | curses.A_BOLD)
        # Score
        stdscr.addstr(0, 2, f"Score: {score}", curses.color_pair(5))
        stdscr.addstr(0, W*2-12, f"Best: {best}", curses.color_pair(5))
        stdscr.addstr(0, W*2-25, f"Lives: {lives}  Level: {level}", curses.color_pair(5))
        if game_over:
            msg = "GAME OVER! Press R to restart, Q to quit"
            stdscr.addstr(H//2+1, (W*2 - len(msg))//2, msg, curses.color_pair(1))
        stdscr.refresh()

    while True:
        key = stdscr.getch()
        if key == ord('q') or key == ord('Q'): break
        if key == ord('r') or key == ord('R'):
            if game_over:
                map_data = [row[:] for row in MAP]
                pac_x, pac_y = 9, 9
                pac_dir = (0,0)
                score = 0
                lives = 3
                level = 1
                game_over = False
                scared_timer = 0
                for g in ghosts:
                    g['x'], g['y'] = g['home']
                    g['dir'] = (0,0)
                    g['scared'] = False
                continue

        if game_over:
            draw()
            continue

        # Управление
        if key == curses.KEY_LEFT or key == ord('a'): pac_dir = (0, -1)
        elif key == curses.KEY_RIGHT or key == ord('d'): pac_dir = (0, 1)
        elif key == curses.KEY_UP or key == ord('w'): pac_dir = (-1, 0)
        elif key == curses.KEY_DOWN or key == ord('s'): pac_dir = (1, 0)

        # Движение Pac-Man
        ny, nx = pac_y + pac_dir[0], pac_x + pac_dir[1]
        if 0 <= ny < H and 0 <= nx < W and map_data[ny][nx] != 0:
            pac_y, pac_x = ny, nx
            # Сбор точек
            if map_data[ny][nx] == 1:
                score += 10
                map_data[ny][nx] = 3
                total_dots -= 1
            elif map_data[ny][nx] == 2:
                score += 50
                map_data[ny][nx] = 3
                total_dots -= 1
                scared_timer = 30  # кадров испуга
                for g in ghosts:
                    g['scared'] = True
                    g['dir'] = (random.choice([-1,0,1]), random.choice([-1,0,1]))

        # Движение призраков
        for g in ghosts:
            if g['scared']:
                # Убегаем от Pac-Man
                dx = g['x'] - pac_x
                dy = g['y'] - pac_y
                # Случайное движение с уклоном
                choices = [(1,0), (-1,0), (0,1), (0,-1)]
                if abs(dx) > abs(dy):
                    choices.sort(key=lambda d: abs(d[0]) if d[0]!=0 else 0, reverse=True)
                else:
                    choices.sort(key=lambda d: abs(d[1]) if d[1]!=0 else 0, reverse=True)
                for d in choices:
                    nx2 = g['x'] + d[0]
                    ny2 = g['y'] + d[1]
                    if 0 <= nx2 < W and 0 <= ny2 < H and map_data[ny2][nx2] != 0:
                        g['x'], g['y'] = nx2, ny2
                        break
            else:
                # Простое случайное движение
                dirs = [(1,0), (-1,0), (0,1), (0,-1)]
                random.shuffle(dirs)
                for d in dirs:
                    nx2 = g['x'] + d[0]
                    ny2 = g['y'] + d[1]
                    if 0 <= nx2 < W and 0 <= ny2 < H and map_data[ny2][nx2] != 0:
                        g['x'], g['y'] = nx2, ny2
                        break

        # Проверка столкновений с призраками
        for g in ghosts:
            if g['x'] == pac_x and g['y'] == pac_y:
                if g['scared']:
                    # Съедаем призрака
                    score += 100
                    g['x'], g['y'] = g['home']
                    g['scared'] = False
                    g['dir'] = (0,0)
                else:
                    lives -= 1
                    if lives <= 0:
                        game_over = True
                        if score > best:
                            best = score
                            save_record(best)
                    else:
                        # Респавн
                        pac_x, pac_y = 9, 9
                        for g2 in ghosts:
                            g2['x'], g2['y'] = g2['home']
                            g2['scared'] = False
                break

        if scared_timer > 0:
            scared_timer -= 1
            if scared_timer == 0:
                for g in ghosts:
                    g['scared'] = False

        # Проверка завершения уровня
        if total_dots == 0:
            level += 1
            map_data = [row[:] for row in MAP]
            total_dots = sum(row.count(1) + row.count(2) for row in map_data)
            pac_x, pac_y = 9, 9
            for g in ghosts:
                g['x'], g['y'] = g['home']
                g['scared'] = False
                g['dir'] = (0,0)

        draw()
        time.sleep(frame)

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('-s', '--speed', type=int, default=150, help='Скорость (мс)')
    args = parser.parse_args()
    try:
        curses.wrapper(main, args.speed)
    except KeyboardInterrupt:
        print("\nИгра завершена.")
